File: gateway/src/iface/authcontrol/authority_client.go
package authcontrol

import (
	"context"
	"time"

	authmodel "gateway/src/models/auth"
)

// DownstreamGrantRequest 定义网关申请下游授权请求。
type DownstreamGrantRequest struct {
	Identity authmodel.IdentityContext

	TargetService string
	TTLSec        int64
}

// UserPasswordAuthRequest 定义用户名密码认证请求。
type UserPasswordAuthRequest struct {
	Username string
	Password string

	Audience string
	Scopes   []string

	ClientID  string
	GatewayID string
	SourceIP  string
	UserAgent string

	RequestID string
	TraceID   string
}

// UserPasswordAuthResult 定义用户名密码认证结果。
type UserPasswordAuthResult struct {
	Identity *authmodel.IdentityContext
	Session  *authmodel.Session
	Tokens   authmodel.TokenBundle

	IssuedAt  time.Time
	ExpiresAt time.Time
}

// IAuthAuthorityClient 定义网关侧对认证中心的客户端调用。
type IAuthAuthorityClient interface {
	InitBootstrapChallenge(ctx context.Context, req *authmodel.ChallengeRequest) (*authmodel.ChallengePayload, error)
	AuthenticateBootstrap(ctx context.Context, req *authmodel.BootstrapAuthRequest) (*authmodel.BootstrapAuthResult, error)

	VerifyToken(ctx context.Context, req *TokenVerifyRequest) (*authmodel.TokenVerificationResult, error)
	ValidateSession(ctx context.Context, req *SessionValidateRequest) (*authmodel.Session, error)
	IssueDownstreamGrant(ctx context.Context, req *DownstreamGrantRequest) (*authmodel.DownstreamAccessGrant, error)

	AuthenticateUserPassword(ctx context.Context, req *UserPasswordAuthRequest) (*UserPasswordAuthResult, error)
	RefreshTokenBundle(ctx context.Context, req *TokenRefreshRequest) (*authmodel.TokenBundle, error)
	RevokeToken(ctx context.Context, req *TokenRevokeRequest) error
}

File: data_server/src/iface/auth/authority_client.py
from abc import ABC, abstractmethod
from uuid import UUID

from msgspec import Struct

from src.models.auth.auth import TokenBundle
from src.models.auth.bootstrap import (
    BootstrapAuthRequest,
    BootstrapAuthResult,
    ChallengePayload,
    ChallengeRequest,
)


class TokenRevokeRequest(Struct, kw_only=True):
    """表示一个令牌撤销请求，包括令牌ID、家族ID和会话ID，以及撤销原因和撤销者等信息。"""

    token_id: UUID
    family_id: UUID
    session_id: UUID

    reason: str
    revoked_by: str

    request_id: str
    trace_id: str


class TokenRefreshRequest(Struct, kw_only=True):
    """表示一个令牌刷新请求，包括当前的访问令牌和刷新令牌，以及客户端信息和请求上下文等。"""

    refresh_token: str

    client_id: str
    gateway_id: str
    source_ip: str
    user_agent: str

    request_id: str
    trace_id: str


class IAuthAuthorityClient(ABC):
    """data_server 侧对认证中心的客户端调用。"""

    @abstractmethod
    async def init_bootstrap_challenge(self, req: ChallengeRequest) -> ChallengePayload:
        raise NotImplementedError

    @abstractmethod
    async def authenticate_bootstrap(
        self, req: BootstrapAuthRequest
    ) -> BootstrapAuthResult:
        raise NotImplementedError

    @abstractmethod
    async def refresh_token_bundle(self, req: TokenRefreshRequest) -> TokenBundle:
        raise NotImplementedError

    @abstractmethod
    async def revoke_token(self, req: TokenRevokeRequest) -> None:
        raise NotImplementedError
